File: cpp/cryptolab/src/rsa_diy.cpp
#include "cryptolab/rsa_diy.hpp"

#include <limits>
#include <stdexcept>
#include <vector>

#include <openssl/rand.h>

namespace cryptolab::rsa_diy {
namespace {

mpz_class from_unsigned(std::uint64_t value) {
  mpz_class result;
  mpz_import(result.get_mpz_t(), 1, 1, sizeof(value), 0, 0, &value);
  return result;
}

mpz_class generate_prime(std::size_t bits, std::uint64_t e) {
  if (bits < 2) {
    throw std::invalid_argument{"prime size must be at least 2 bits"};
  }

  std::vector<unsigned char> random_bytes((bits + 7) / 8);
  const auto exponent = from_unsigned(e);

  for (;;) {
    if (RAND_bytes(random_bytes.data(), random_bytes.size()) != 1) {
      throw std::runtime_error{"OpenSSL random-number generation failed"};
    }

    mpz_class candidate;
    mpz_import(candidate.get_mpz_t(), random_bytes.size(), 1,
               sizeof(random_bytes.front()), 1, 0, random_bytes.data());
    mpz_fdiv_r_2exp(candidate.get_mpz_t(), candidate.get_mpz_t(), bits);
    mpz_setbit(candidate.get_mpz_t(), bits - 1);
    mpz_setbit(candidate.get_mpz_t(), bits - 2);
    mpz_setbit(candidate.get_mpz_t(), 0);

    mpz_class divisor;
    const mpz_class candidate_minus_one = candidate - 1;
    mpz_gcd(divisor.get_mpz_t(), exponent.get_mpz_t(),
            candidate_minus_one.get_mpz_t());
    if (divisor != 1) {
      continue;
    }

    if (mpz_probab_prime_p(candidate.get_mpz_t(), 40) > 0) {
      return candidate;
    }
  }
}

} // namespace

KeyPair::KeyPair(std::uint64_t e, std::size_t modulus_bits) : e_{e} {
  if (e <= 1 || e % 2 == 0) {
    throw std::invalid_argument{"e must be an odd integer greater than 1"};
  }
  if (modulus_bits < 16) {
    throw std::invalid_argument{"modulus size must be at least 16 bits"};
  }

  const auto p_bits = modulus_bits / 2;
  const auto q_bits = modulus_bits - p_bits;
  const auto p = generate_prime(p_bits, e_);

  mpz_class q;
  do {
    q = generate_prime(q_bits, e_);
  } while (q == p);

  modulus_ = p * q;
  const mpz_class phi = (p - 1) * (q - 1);
  const auto exponent = from_unsigned(e_);
  if (mpz_invert(d_.get_mpz_t(), exponent.get_mpz_t(), phi.get_mpz_t()) == 0) {
    throw std::runtime_error{"public exponent is not invertible modulo phi(N)"};
  }
}

std::pair<std::uint64_t, mpz_class> KeyPair::public_key() const {
  return {e_, modulus_};
}

std::pair<mpz_class, mpz_class> KeyPair::private_key() const {
  return {d_, modulus_};
}

mpz_class KeyPair::encrypt(std::uint64_t plaintext) const {
  mpz_class ciphertext;
  const auto message = from_unsigned(plaintext);
  const auto exponent = from_unsigned(e_);
  mpz_powm(ciphertext.get_mpz_t(), message.get_mpz_t(), exponent.get_mpz_t(),
           modulus_.get_mpz_t());
  return ciphertext;
}

std::uint64_t KeyPair::decrypt(const mpz_class &ciphertext) const {
  mpz_class plaintext;
  mpz_powm(plaintext.get_mpz_t(), ciphertext.get_mpz_t(), d_.get_mpz_t(),
           modulus_.get_mpz_t());
  if (plaintext < 0 ||
      plaintext > from_unsigned(std::numeric_limits<std::uint64_t>::max())) {
    throw std::overflow_error{"decrypted plaintext does not fit in uint64_t"};
  }

  std::uint64_t result = 0;
  std::size_t count = 0;
  mpz_export(&result, &count, 1, sizeof(result), 0, 0, plaintext.get_mpz_t());
  return result;
}

} // namespace cryptolab::rsa_diy

File: rs/cryptolab/src/rsa_diy.rs
use rand::{RngCore, rngs::OsRng};
use rug::{Complete, Integer, integer::IsPrime, integer::Order};

pub const DEFAULT_PUBLIC_EXPONENT: u64 = 65_537;
pub const DEFAULT_MODULUS_BITS: u32 = 4_096;

pub struct KeyPair {
    e: u64,
    modulus: Integer,
    d: Integer,
}

impl KeyPair {
    pub fn new(e: u64, modulus_bits: u32) -> Result<Self, String> {
        if e <= 1 || e.is_multiple_of(2) {
            return Err("e must be an odd integer greater than 1".into());
        }
        if modulus_bits < 16 {
            return Err("modulus size must be at least 16 bits".into());
        }

        let p_bits = modulus_bits / 2;
        let q_bits = modulus_bits - p_bits;
        let p = generate_prime(p_bits, e)?;
        let q = loop {
            let candidate = generate_prime(q_bits, e)?;
            if candidate != p {
                break candidate;
            }
        };

        let modulus = Integer::from(&p * &q);
        let phi = Integer::from(&p - 1) * Integer::from(&q - 1);
        let d = Integer::from(e)
            .invert(&phi)
            .map_err(|_| "public exponent is not invertible modulo phi(N)".to_owned())?;

        Ok(Self { e, modulus, d })
    }

    pub fn default_rsa() -> Result<Self, String> {
        Self::new(DEFAULT_PUBLIC_EXPONENT, DEFAULT_MODULUS_BITS)
    }

    pub fn public_key(&self) -> (u64, &Integer) {
        (self.e, &self.modulus)
    }

    pub fn private_key(&self) -> (&Integer, &Integer) {
        (&self.d, &self.modulus)
    }

    pub fn encrypt(&self, plaintext: u64) -> Result<Integer, String> {
        Integer::from(plaintext)
            .pow_mod(&Integer::from(self.e), &self.modulus)
            .map_err(|_| "DIY RSA encryption failed".to_owned())
    }

    pub fn decrypt(&self, ciphertext: &Integer) -> Result<u64, String> {
        ciphertext
            .clone()
            .pow_mod(&self.d, &self.modulus)
            .map_err(|_| "DIY RSA decryption failed".to_owned())?
            .to_u64()
            .ok_or_else(|| "decrypted plaintext does not fit in u64".to_owned())
    }
}

fn generate_prime(bits: u32, e: u64) -> Result<Integer, String> {
    if bits < 2 {
        return Err("prime size must be at least 2 bits".into());
    }

    let mut random_bytes = vec![0_u8; bits.div_ceil(8) as usize];
    let exponent = Integer::from(e);

    loop {
        OsRng
            .try_fill_bytes(&mut random_bytes)
            .map_err(|error| format!("OS random-number generation failed: {error}"))?;

        let mut candidate = Integer::from_digits(&random_bytes, Order::MsfBe);
        candidate.keep_bits_mut(bits);
        candidate.set_bit(bits - 1, true);
        candidate.set_bit(bits - 2, true);
        candidate.set_bit(0, true);

        let mut candidate_minus_one = candidate.clone();
        candidate_minus_one -= 1;
        if candidate_minus_one.gcd_ref(&exponent).complete() != 1 {
            continue;
        }

        if candidate.is_probably_prime(40) != IsPrime::No {
            return Ok(candidate);
        }
    }
}
